From 464422f95004803e96fcf8c7000a33d743fa11cd Mon Sep 17 00:00:00 2001
From: Sandor Yu <R01008@freescale.com>
Date: Fri, 30 Jan 2015 17:25:40 +0800
Subject: [PATCH 0824/1594] MLK-10182: cam :Module can't load/unload twice
 without camera connected

commit ca9fef6754ea87c1faa0f8ce7dc75320fb4f3de5 from
git://git.freescale.com/imx/linux-2.6-imx.git

If no camera connected, ov5640 driver loading failed,
but the driver have register subdev by v4l2_async_register_subdev.
v4l2_async_unregister_subdev function in remove function
and not been called when module unloading.
Same subdev can not register twice, v4l2 async driver will print
error message in the second module loading.
Move v4l2_async_register_subdev function after ov5640 is succeed
found to fix the issue.

Signed-off-by: Sandor Yu <R01008@freescale.com>
---
 drivers/media/platform/mxc/subdev/ov5640.c |   14 +++++++-------
 1 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/media/platform/mxc/subdev/ov5640.c b/drivers/media/platform/mxc/subdev/ov5640.c
index 9ca3d8a..6a7586f 100644
--- a/drivers/media/platform/mxc/subdev/ov5640.c
+++ b/drivers/media/platform/mxc/subdev/ov5640.c
@@ -1834,13 +1834,6 @@ static int ov5640_probe(struct i2c_client *client,
 		return retval;
 	}
 
-	v4l2_i2c_subdev_init(&ov5640_data.subdev, client, &ov5640_subdev_ops);
-
-	retval = v4l2_async_register_subdev(&ov5640_data.subdev);
-	if (retval < 0)
-		dev_err(&client->dev,
-					"%s--Async register faialed, ret=%d\n", __func__, retval);
-
 	clk_prepare_enable(ov5640_data.sensor_clk);
 
 	ov5640_data.io_init = ov5640_reset;
@@ -1885,6 +1878,13 @@ static int ov5640_probe(struct i2c_client *client,
 
 	clk_disable(ov5640_data.sensor_clk);
 
+	v4l2_i2c_subdev_init(&ov5640_data.subdev, client, &ov5640_subdev_ops);
+
+	retval = v4l2_async_register_subdev(&ov5640_data.subdev);
+	if (retval < 0)
+		dev_err(&client->dev,
+					"%s--Async register failed, ret=%d\n", __func__, retval);
+
 	pr_info("camera ov5640, is found\n");
 	return retval;
 }
-- 
1.7.5.4

